tests/test_grad.py

```diff
@@ -67,4 +73,7 @@ def test_example_7_m():
     ])
     memory_grad = do_grad(spec, pi, 'm', lr=1)
 
-    assert np.allclose(memory_end, memory_grad, atol=1e-5)
+    assert np.allclose(memory_end, softmax(memory_grad, axis=-1), atol=1e-2)
```

Comment:
I'm a little nervous about using such a large tolerance. Are we sure it's doing what we want if we allow it to be off by 1e-2? Some of these value functions vary by smaller amounts than that.

Reply:
For value functions, I would totally agree - yes, a tolerance of 1e-2 is too large, seeing as the scale of the returns will depend on reward sparsity and discount rate.

But since we're essentially comparing two probability distributions which are normalized between 0 and 1, I'm more comfortable with this tolerance.

I've also looked into trying to make this tolerance smaller, and the approaches I've tried haven't done what I expected them to. I've tried low-temperature softmax, scaling the learning rate, scaling the loss - all seem to result in similar tolerances. It seems that the softmax function is quite stubborn in terms of getting large enough parameters for a small tolerance.

Reply:
Fair enough. I just didn't want it to end up biting us if we thought it was converged because the test passed, but really it's off by some significant amount that's still within the large tolerance.